### ISSUE TITLE
OperationCorrelationTelemetryInitializer correlating unexpected requests

### ISSUE THREAD

**Issue by GeorgeTsiokos** [Microsoft/ApplicationInsights-dotnet#277]
Given this _Hello World_ **RabbitMQ + ApplicationInsights** sample, I'm receiving unexpected parent operation ids:
https://gist.github.com/GeorgeTsiokos/3cd921fc51b35c188f3b02addd93a143

```
Press <enter> to stop
 [x] Sent Hello World!          Telemetry Operation = BasicPublish
 [x] Received Hello World!      Telemetry Operation = Startup
 [x] Sent Hello World!          Telemetry Operation = BasicPublish
 [x] Received Hello World!      Telemetry Operation = Startup

Press <enter> to quit
```

_the completed startup event is correlated will all messages received from RabbitMQ_

Timeline of events:
- Startup Begin
  - PublisherStartup
  - SubscriberStartup
- Startup End
- ConsoleLoop Begin
  - BasicPublish
  - Received
- ConsoleLoop End

If you remove the call to `ExecutionContext.SuppressFlow()`, the **Sent Hello World!** will display a telemetry operation of **Startup** as well, which is definitely not a typical use case, especially if you want to create a Application Insights "Request" for every Rabbit MQ message received. For some unknown reason, moving the creation of the `factory`, `connection`, and `channel` outside of the **Startup** request resolves the RabbitMQ problem.

Two questions:
- When it comes to delegates and events, do you have guidance as to how to instantiate 3rd party libraries / subscribe to asynchronous events?
- Is it possible to add proactive detection of incorrect telemetry by detecting completed operations and observing new telemetry correlated with the completed telemetry?

**Comment by SergeyKanzhelev** [comment]
@GeorgeTsiokos, sorry for delay. 

So you see that dispose of startup event doesn't clean up context as expected: https://github.com/Microsoft/ApplicationInsights-dotnet/blob/1ccd5b466451db1884b242b6a58b3e9401a1fbcf/src/Core/Managed/Operation.CC.Shared/Extensibility/Implementation/CallContextHelpers.cs#L36 or https://github.com/Microsoft/ApplicationInsights-dotnet/blob/1ccd5b466451db1884b242b6a58b3e9401a1fbcf/src/Core/Managed/Operation.AL.Shared/Extensibility/Implementation/AsyncLocalHelpers.cs#L32

I've created this unit test and it passes all right. Do you know whether the factory/channel do anything fanky with call context? 

``` csharp
        /// <summary>
        /// Ensure that context being propagated via async/await.
        /// </summary>
        [TestMethod]
        public void ContextCleansUp()
        {
            var task = this.TestContextCleansUpAfterDisposeAsync();
            task.Wait();
        }

        /// <summary>
        /// Actual async test method.
        /// </summary>
        /// <returns>Task to await.</returns>
        public async Task TestContextCleansUpAfterDisposeAsync()
        {
            using (var op = this.telemetryClient.StartOperation<RequestTelemetry>("request"))
            {
                var id1 = Thread.CurrentThread.ManagedThreadId;
                this.telemetryClient.TrackTrace("trace1");

                //HttpClient client = new HttpClient();
                await Task.Delay(100);//client.GetStringAsync("http://bing.com");

                var id2 = Thread.CurrentThread.ManagedThreadId;
                this.telemetryClient.TrackTrace("trace2");

                Assert.NotEqual(id1, id2);
            }

            using (var op = this.telemetryClient.StartOperation<RequestTelemetry>("request2"))
            {
            }


            Assert.Equal(4, this.sendItems.Count);
            var request2 = ((RequestTelemetry)this.sendItems[this.sendItems.Count - 1]);
            Assert.Equal("request2", request2.Name);
            Assert.False(string.IsNullOrEmpty(request2.Id));
            Assert.Equal(request2.Id, request2.Context.Operation.Id);

        }
```


We have a warning trace when you attempt to stop an incorrect operation: https://github.com/Microsoft/ApplicationInsights-dotnet/blob/1ccd5b466451db1884b242b6a58b3e9401a1fbcf/src/Core/Managed/Operation.CC.Shared/Extensibility/Implementation/CallContextBasedOperationHolder.cs#L78 To enable this warning - change severity on `DiagnosticsTelemetryModule` using `ApplicationInsights.config` and setting this property: https://github.com/Microsoft/ApplicationInsights-dotnet/blob/0110215df15bca47aa7fc7ee92c46f30d909dead/src/Core/Managed/Shared/Extensibility/Implementation/Tracing/DiagnosticsTelemetryModule.cs#L56

It is quite hard to implement warning you are asking for as we do not store operations after they were sent. Do you have a suggestion how we can implement it?

**Comment by GeorgeTsiokos** [comment]
Here's a simpler repro where the `System.Threading.Timer` class constructor captures the ambient application insights _request_ on object creation and never lets it go, even though the application insights request completed, and, imo has nothing to do with the timer callback:
```c#
var telemetryClient = new TelemetryClient();

TimerCallback callExternalService = _ =>
{
    using (telemetryClient.StartOperation<DependencyTelemetry>("Call external service"))
    {
        Thread.Sleep(500);
    }
};

Timer asyncWorker;
using (telemetryClient.StartOperation<RequestTelemetry>("StartService"))
{
    asyncWorker = new Timer(callExternalService);
}

asyncWorker.Change(1000, 1000);
```
Here's the output (notice __StartService__ corrilating with the timer):
```JSON
{
  "name": "Microsoft.ApplicationInsights.Dev.RemoteDependency",
  "time": "2016-08-24T15:28:35.2643013Z",
  "tags": {
    "ai.operation.parentId": "S/GrcRNTa/k=",
    "ai.operation.name": "StartService",
    "ai.operation.id": "S/GrcRNTa/k=",
    "ai.internal.sdkVersion": "dotnet: 2.1.0.26048"
  },
  "data": {
    "baseType": "RemoteDependencyData",
    "baseData": {
      "ver": 2,
      "name": "Call external service",
      "id": "7ZFAC3aUDhg=",
      "value": 500.6487,
      "dependencyKind": 2,
      "success": true,
      "properties": { "DeveloperMode": "true" }
    }
  }
}
```
Here's the completed request _(the first AI entry)_:
```JSON
{
  "name": "Microsoft.ApplicationInsights.Dev.Request",
  "time": "2016-08-24T15:28:21.2425565Z",
  "tags": {
    "ai.operation.name": "StartService",
    "ai.operation.id": "S/GrcRNTa/k=",
    "ai.internal.sdkVersion": "dotnet: 2.1.0.26048"
  },
  "data": {
    "baseType": "RequestData",
    "baseData": {
      "ver": 2,
      "id": "S/GrcRNTa/k=",
      "name": "StartService",
      "startTime": "2016-08-24T11:28:21.2425565-04:00",
      "duration": "00:00:00.0039559",
      "success": true,
      "responseCode": "200",
      "properties": { "DeveloperMode": "true" }
    }
  }
}
```

**Comment by SergeyKanzhelev** [comment]
@GeorgeTsiokos - this code would behave the same way - Application Insights will correlate Request `request` and two dependency calls - one to `bing.com` and another to `google.com`. And it doesn't matter whether the dependency calls will complete long after request has already been completed.

``` csharp
using (this.telemetryClient.StartOperation<RequestTelemetry>("request"))
{
  var id1 = Thread.CurrentThread.ManagedThreadId;
  this.telemetryClient.TrackTrace("trace1");

  HttpClient client = new HttpClient();
  Task t1 = client.GetStringAsync("http://bing.com");
  Task t2 = client.GetStringAsync("http://google.com");
}
```
This is a desired behavior.

Is your question how to "clear up" the operation context in timer callback?

**Comment by GeorgeTsiokos** [comment]
I agree in this case you would expect the HTTP requests to be correlated with the AI _request_. For comparison, moving the constrictor of the `HttpClient` to a separate request had no impact on the behavior:
```c#
HttpClient client;
using (telemetryClient.StartOperation<RequestTelemetry>("init"))
{
    client = new HttpClient();
}
```
For another comparison, I used `System.Timers.Timer`:
```c#
ElapsedEventHandler elapsedEventHandler = (sender, eventArgs) =>
{
    using (telemetryClient.StartOperation<DependencyTelemetry>("Call external service"))
    {
        Thread.Sleep(500);
    }
};

Timer asyncWorker;
using (telemetryClient.StartOperation<RequestTelemetry>("StartService"))
{
    asyncWorker = new Timer(1000);
    asyncWorker.Elapsed += elapsedEventHandler;
    Thread.Sleep(500);
}

asyncWorker.Start();
```
This timer works as expected - the elapsed event is not correlated with the _StartService_ request. I found moving the call to `Start` controls when the ambient AI _request_ is captured. For `System.Threading.Timer`, is appears that the ambient AI request is captured in the constructor.

I guess my problem here is that the external types I'm using do not document when they capture the ambient `ExecutionContext` which leads to unexpected AI event correlations...

To suppress this behavior, I've found that decorating `using (ExecutionContext.SuppressFlow())` resolves this issue.
